lib/Sema/MiscDiagnostics.cpp

```diff
@@ -1297,7 +1300,7 @@ static void diagSyntacticUseRestrictions(const Expr *E, const DeclContext *DC,
       if (dyn_cast<NilLiteralExpr>(E)) return true;
 
       auto CE = dyn_cast<ApplyExpr>(E->getSemanticsProvidingExpr());
-      if (!CE || !CE->isImplicit())
```

Comment:
I think it would be better to actually check for `Optional.none` this way:

```
if (auto *DSCE = dyn_cast_or_null<DotSyntaxCallExpr>(E->getSemanticsProvidingExpr())) {
  if (auto *DRE = dyn_cast<DeclRefExpr>(DSCE->getSemanticFn()))
    return DRE->getDecl() == Ctx.getOptionalNoneDecl();
}
```

This lets us to avoid null and implicit checks.

Reply:
@xedin, thanks for the suggestion -- updated in https://github.com/apple/swift/pull/60893/commits/d9be4a721311dae452f89f3e74ca332855c3ba99